File: worker/src/sequenceLoader.ts
import * as fs from "fs"
import * as path from "path"
import * as yaml from "js-yaml"

// Read from env if it exists, otherwise default to local relative path
const SEQUENCES_DIR = process.env.SEQUENCES_DIR ?? path.resolve(__dirname, "../../sequences")

const cache: Record<string, any> = {}

/**
 * Loads a sequence YAML file by ID.
 * Cached in memory after first load.
 */
export function loadSequence(sequenceId: string): any | null {
  if (cache[sequenceId]) return cache[sequenceId]

  const filePath = path.join(SEQUENCES_DIR, `${sequenceId}.yaml`)
  if (!fs.existsSync(filePath)) return null

  const data = yaml.load(fs.readFileSync(filePath, "utf8"))
  cache[sequenceId] = data
  return data
}

/**
 * Finds a specific step inside a sequence by its ID.
 */
export function findStep(sequence: any, stepId: string): any | null {
  return sequence.steps?.find((s: any) => s.id === stepId) ?? null
}

File: apps/api/src/services/sequence_loader.py
import yaml
import os
from typing import Optional

# Check for an environment variable first, default to local relative path
SEQUENCES_DIR = os.getenv("SEQUENCES_DIR", os.path.join(
    os.path.dirname(__file__), "../../../../sequences"
))

_cache: dict = {}


def load_sequence(sequence_id: str) -> Optional[dict]:
    """
    Loads a sequence definition from a YAML file.
    Caches it in memory after first load so we don't hit disk on every request.

    Expected file path:
        sequences/{sequence_id}.yaml

    Returns None if the sequence file doesn't exist.
    """
    if sequence_id in _cache:
        return _cache[sequence_id]

    path = os.path.join(SEQUENCES_DIR, f"{sequence_id}.yaml")

    if not os.path.exists(path):
        return None

    with open(path, "r") as f:
        data = yaml.safe_load(f)

    _cache[sequence_id] = data
    return data


def parse_delay(delay_str) -> int:
    """
    Converts a human-readable delay string into seconds.

    Examples:
        0       → 0
        "1d"    → 86400
        "12h"   → 43200
        "30m"   → 1800
        "5d"    → 432000
    """
    if delay_str == 0 or delay_str == "0":
        return 0

    delay_str = str(delay_str).strip()

    if delay_str.endswith("d"):
        return int(delay_str[:-1]) * 86400
    elif delay_str.endswith("h"):
        return int(delay_str[:-1]) * 3600
    elif delay_str.endswith("m"):
        return int(delay_str[:-1]) * 60
    else:
        return int(delay_str)